### DISCUSSION TITLE
Custom Script filter for unconnected interfaces

### DISCUSSION THREAD

**Discussion by SRF-rueeggdo** [netbox-community/netbox#17556] [Q&A] [answered]
Hello everyone

I would like to display in a custom script in a field only interfaces that are not yet connected.
Unfortunately I can't do that. Does anyone know how to do this?

I have tried the following:
```
"cable_id__isnull": "true",
"cable_id__isnull": "True,
"cable_id": None,
"cable": "null",
```

Nothing worked

```
    rt_switch_a = ObjectVar(
        label="Device Real Time port A ->",
        description="Wählen Sie aus, an welchen Switch der Realtime Port A Ihres Gerätes angeschlossen werden soll.",
        model=Device,
        required=False,
        query_params={
            "role_id": ["1", "2", "5"],
        },
    )

    rt_switch_port1 = ObjectVar(
        label="Switch Port",
        description="Wählen Sie den Anschluss auf der Switch-Seite.",
        model=Interface,
        required=False,
        query_params={
            "device_id": "$rt_switch_a",
            "cable": "null",
        },

```

**Comment by candlerb** [comment, marked as answer]
TL;DR: try `"occupied": False`

Looking in the source code, at [InterfaceFilterSet](https://github.com/netbox-community/netbox/blob/v4.1.1/netbox/dcim/filtersets.py#L1652) which includes [CabledObjectFilterSet](https://github.com/netbox-community/netbox/blob/v4.1.1/netbox/dcim/filtersets.py#L1492-L1510) and [PathEndpointFilterSet](https://github.com/netbox-community/netbox/blob/v4.1.1/netbox/dcim/filtersets.py#L1513-L1522), I see these filters:

```
    cabled = django_filters.BooleanFilter(
        field_name='cable',
        lookup_expr='isnull',
        exclude=True
    )
    occupied = django_filters.BooleanFilter(
        method='filter_occupied'
    )

    def filter_occupied(self, queryset, name, value):
        if value:
            return queryset.filter(Q(cable__isnull=False) | Q(mark_connected=True))
        else:
            return queryset.filter(cable__isnull=True, mark_connected=False)

...

    connected = django_filters.BooleanFilter(
        method='filter_connected'
    )

    def filter_connected(self, queryset, name, value):
        if value:
            return queryset.filter(_path__is_active=True)
        else:
            return queryset.filter(Q(_path__isnull=True) | Q(_path__is_active=False))
```

That is, I believe the filters you could use are `cabled=False`, `occupied=False`, `connected=False`.

These are the same filters you'll find in the web UI if you go to `/dcim/interfaces/`, click on the Filter tab, and scroll to the bottom:

![image](https://github.com/user-attachments/assets/6d30e033-d220-4fcf-80a9-a896a64752e2)

As far as I can tell, the differences are:

* "cabled" means a cable is present in the interface
* "occupied" means that either a cable is present in the interface, or the interface is marked "connected"
* "connected" means that a cable is present and also forms part of a complete path all the way to another interface. That is, it will exclude when a cable is inserted but it only connects to a patch panel and no further.

In your case, I think `occupied=False` is what you want: it will then only offer interfaces which it makes sense to add a cable to.

**Reply by zmaster7** [reply]
Edit: Scratch that, it seems to still not filter properly... oops!
Edit2: As suggested above, ["occupied": False] does work.

If the solution provided by @candlerb does not work, you can utilized "connected_endpoints": [], I just tested it. Since an interface can connect to multiple endpoints, a null value would be an empty array.

```
device = ObjectVar(
        label = "Device/Swith",
        model = Device,
        required = False
    )

    unusedInterface = ObjectVar(
        label = "Switch Port",
        model = Interface,
        required = False,
        query_params={
            "device_id": "$device",
            "connected_endpoints": [],
        },
    )
```

This can be confirmed relatively easily via:

```
def run(self, data, commit):
        interface = data['unusedInterface']

        self.log_info(f'Device {interface.device.name} - Interface {interface.name} - Connected Endpoints {len(interface.connected_endpoints)}')
```

**Reply by candlerb** [reply]
To be used as `query_params` in an `ObjectVar`, it has to be something which is usable as an API filter (a URL query string), because it's being used to build a URL for Javascript to call out to. Just because the Interface model has a property "connected_endpoints" doesn't mean it's exposed as a URL query.

I believe the same filters are available in the GUI URLs as well as REST URLs, and a quick test shows

`/dcim/interfaces/?connected_endpoints=`
`/dcim/interfaces/?connected_endpoints=[]`

both return the full set of interfaces (Netbox doesn't reject invalid query parameters, it just ignores them).
